core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -536,10 +556,22 @@ class ReplicaManager(val config: KafkaConfig,
            * This can cause a replica to always be out of sync.
            */
           val initialLogEndOffset = localReplica.logEndOffset
+
           val logReadInfo = localReplica.log match {
             case Some(log) =>
               val adjustedFetchSize = if (Topic.isInternal(topic) && !readOnlyCommitted) Math.max(fetchSize, log.config.maxMessageSize) else fetchSize
-              log.read(offset, adjustedFetchSize, maxOffsetOpt)
+
+              //Try the read first, this tells us whether we need all of adjustedFetchSize for this partition
+              var fetch = log.read(offset, adjustedFetchSize, maxOffsetOpt)
+
+              //If the read for this partition caused the quota to be excedded then zero it out, so it is excluded
+              if(quota.isThrottled(TopicAndPartition(topic, partition))) {
+                if (quota.isQuotaExceededBy(check(totalThrottledBytesInFetch, quota.bound())))
```

Comment:
Sorry - deleted previous. As I understand it the fetch, here, is just a single partition's data. Thus I'm keeping a running total of throttled bytes in all partitions, checking against the quota to see when the 'proposed' fetch will exceed ot. Remember isQuotaExceededBy() doesn't record anything. Let me know if I'm missing something.